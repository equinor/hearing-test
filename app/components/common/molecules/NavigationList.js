import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  View,
  StyleSheet,
  Text,
  ScrollView,
  FlatList,
  TouchableOpacity,
  RefreshControl,
} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import * as colors from '../../../stylesheets/colors';

const styles = StyleSheet.create({
  scrollView: {
    flex: 1,
    backgroundColor: colors.GRAY_BACKGROUND,
  },
  itemContainer: {
    backgroundColor: 'white',
    height: 56,
    borderBottomWidth: 1,
    borderBottomColor: colors.GRAY_3,
    flexDirection: 'row',
    alignItems: 'center',
  },
  icon: {
    color: colors.BLACK_LIGHT,
    alignSelf: 'center',
  },
  textStyle: {
    fontSize: 16,
    color: colors.BLACK_GRAY,
    flex: 1,
    paddingHorizontal: 15,
  },
});

export default class NavigationList extends Component {
  static propTypes = {
    items: PropTypes.arrayOf(
      PropTypes.shape({
        key: PropTypes.string,
        label: PropTypes.string,
        route: PropTypes.string,
        params: PropTypes.object,
      })
    ).isRequired,
    fetching: PropTypes.bool,
    navigation: PropTypes.object,
    onRefresh: PropTypes.func,
  };

  static defaultProps = {
    fetching: false,
    navigation: null,
    onRefresh: null,
  };

  handleClick = index => {
    const item = this.props.items[index];
    this.props.navigation.navigate(item.route, item.params);
  };

  renderItem = ({ item, index }) => (
    <TouchableOpacity onPress={() => this.handleClick(index)} underlayColor={colors.GRAY_3}>
      <View style={styles.itemContainer}>
        <Text style={styles.textStyle}>{item.label}</Text>
        <Icon name="chevron-right" style={styles.icon} size={26} />
      </View>
    </TouchableOpacity>
  );

  render() {
    const { items, onRefresh, fetching } = this.props;
    return (
      <ScrollView
        style={styles.scrollView}
        refreshControl={onRefresh && <RefreshControl refreshing={fetching} onRefresh={onRefresh} />}
      >
        <FlatList data={items} keyExtractor={item => item.key} renderItem={this.renderItem} />
      </ScrollView>
    );
  }
}
