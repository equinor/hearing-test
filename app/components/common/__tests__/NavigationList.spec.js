import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import NavigationList from '../molecules/NavigationList';

it('renders and behaves correctly', () => {
  let navInfo;

  const items = [
    {
      key: 'item1Key',
      label: 'item1Label',
      route: 'item1Route',
    },
    {
      key: 'item2Key',
      label: 'item2Label',
      route: 'item2Route',
      params: { id: 2 },
    },
  ];
  const nav = {
    navigate: sinon.spy((route, params) => {
      navInfo = { route, params };
    }),
  };

  const onRefresh = sinon.spy();

  const wrapper = shallow(<NavigationList items={items} navigation={nav} onRefresh={onRefresh} />);
  const instance = wrapper.instance();
  expect(wrapper).toMatchSnapshot();
  expect(nav.navigate.calledOnce).toBeFalsy();
  instance.handleClick(0);
  expect(nav.navigate.calledOnce).toBeTruthy();
  expect(navInfo.route).toBe('item1Route');
  expect(navInfo.params).toBeUndefined();
  instance.handleClick(1);
  expect(nav.navigate.calledTwice).toBeTruthy();
  expect(navInfo.route).toBe('item2Route');
  expect(navInfo.params).toEqual({ id: 2 });
});
