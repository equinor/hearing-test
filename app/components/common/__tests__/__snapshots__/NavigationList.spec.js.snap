// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and behaves correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NavigationList
    fetching={false}
    items={
      Array [
        Object {
          "key": "item1Key",
          "label": "item1Label",
          "route": "item1Route",
        },
        Object {
          "key": "item2Key",
          "label": "item2Label",
          "params": Object {
            "id": 2,
          },
          "route": "item2Route",
        },
      ]
    }
    navigation={
      Object {
        "navigate": [Function],
      }
    }
    onRefresh={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <FlatList
        data={
          Array [
            Object {
              "key": "item1Key",
              "label": "item1Label",
              "route": "item1Route",
            },
            Object {
              "key": "item2Key",
              "label": "item2Label",
              "params": Object {
                "id": 2,
              },
              "route": "item2Route",
            },
          ]
        }
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        numColumns={1}
        onEndReachedThreshold={2}
        renderItem={[Function]}
        scrollEventThrottle={50}
        updateCellsBatchingPeriod={50}
        windowSize={21}
      />,
      "refreshControl": <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />,
      "style": Object {
        "backgroundColor": undefined,
        "flex": 1,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "data": Array [
          Object {
            "key": "item1Key",
            "label": "item1Label",
            "route": "item1Route",
          },
          Object {
            "key": "item2Key",
            "label": "item2Label",
            "params": Object {
              "id": 2,
            },
            "route": "item2Route",
          },
        ],
        "disableVirtualization": false,
        "horizontal": false,
        "initialNumToRender": 10,
        "keyExtractor": [Function],
        "maxToRenderPerBatch": 10,
        "numColumns": 1,
        "onEndReachedThreshold": 2,
        "renderItem": [Function],
        "scrollEventThrottle": 50,
        "updateCellsBatchingPeriod": 50,
        "windowSize": 21,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [MockFunction],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <FlatList
          data={
            Array [
              Object {
                "key": "item1Key",
                "label": "item1Label",
                "route": "item1Route",
              },
              Object {
                "key": "item2Key",
                "label": "item2Label",
                "params": Object {
                  "id": 2,
                },
                "route": "item2Route",
              },
            ]
          }
          disableVirtualization={false}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          numColumns={1}
          onEndReachedThreshold={2}
          renderItem={[Function]}
          scrollEventThrottle={50}
          updateCellsBatchingPeriod={50}
          windowSize={21}
        />,
        "refreshControl": <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />,
        "style": Object {
          "backgroundColor": undefined,
          "flex": 1,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "data": Array [
            Object {
              "key": "item1Key",
              "label": "item1Label",
              "route": "item1Route",
            },
            Object {
              "key": "item2Key",
              "label": "item2Label",
              "params": Object {
                "id": 2,
              },
              "route": "item2Route",
            },
          ],
          "disableVirtualization": false,
          "horizontal": false,
          "initialNumToRender": 10,
          "keyExtractor": [Function],
          "maxToRenderPerBatch": 10,
          "numColumns": 1,
          "onEndReachedThreshold": 2,
          "renderItem": [Function],
          "scrollEventThrottle": 50,
          "updateCellsBatchingPeriod": 50,
          "windowSize": 21,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [MockFunction],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
